API/PROBLEMS/uncategorized

```diff
@@ -0,0 +1,133 @@
+Which parts of API are present only because the macros we provide to plugin writers:
+- VERNAC COMMAND EXTEND
+- VERNAC ARGUMENT EXTEND
+- TACTIC EXTEND
+- ARGUMENT EXTEND
+expand to a code that needs it.
+
+If there are such entries in the API,
+then it would be wise to move them out to a separate module.
+
+MOTIVATION
+
+- it will be clearer what plugin writer may want to care about
+  (these elements will be in one module)
+  and what he/she does not need to care about
+  (these elements will be in another module)
+
+HINTS
+
+One can see this separation in trunk (e.g. it is still present in commit 35e0f32)
+
+--------------------------------------------------------------------------------
+
+until it was reverted.
+Prelude ... figure out how we could remove this module.
+
+HINTS:
+- https://github.com/coq/coq/pull/808
+  - https://github.com/ejgallego/coq/pull/8
+
+--------------------------------------------------------------------------------
+
+Isn't it odd that the following two functions:
+
+  val it_mkLambda_or_LetIn : constr -> Context.Rel.t -> constr
+  val mkLambda_or_LetIn : rel_declaration -> constr -> constr
+
+are not "similarly general"?
+
+(one can take a context of any kind (Constr as well as EConstr)
+ the other can take only Constr)
+
+?
+
+--------------------------------------------------------------------------------
+
+Does it make sense to keep the elements we have marked as @@ocaml.deprecated ?
+
+--------------------------------------------------------------------------------
+
+API should be properly formatted.
+
+--------------------------------------------------------------------------------
+
+Look at these types
+- Vernacexpr.local_decl_expr
+- Context.{Named,Rel}.Declaration.t
+and check if it would make sense to refactor them
+(define it once as a functor and intantiate just its specializations)
+
+--------------------------------------------------------------------------------
+
+It makes sense to consider to define some recipe how things (types, functions)
+should be ordered in modules (instead of leaving them in random, mutually inconsistent order).
+
+This concern was heeded when Context module was cleaned up
+(things in there are not ordered randomly)
+
+It would make sense to consider extending this idea to all the modules present in the API.
+
+- make sure that it is followed by all modules that are provided by the API
+- update API/PROPERTIES accordingly
+
+--------------------------------------------------------------------------------
+
+What is the plan about the removal of API elements that are currently marked as [@@ocaml.deprecated] ?
```

Comment:
AFAIK these deprecation notices were added in 8.7 with the plan of getting rid of them ASAP (in 8.8).